traitsui/editor.py

```diff
@@ -445,146 +512,121 @@ def sync_value(
-                user_object = self.context_object
-                xuser_name = user_name
-            else:
-                user_object = self.ui.context[user_name[:col]]
-                user_name = xuser_name = user_name[col + 1 :]
-                col = user_name.rfind(".")
-                if col >= 0:
-                    user_ref += "." + user_name[:col]
-                    user_name = user_name[col + 1 :]
-
-            user_value = compile(
-                "%s.%s" % (user_ref, user_name), "<string>", "eval"
-            )
-            user_ref = compile(user_ref, "<string>", "eval")
+        parts = user_name.split(".")
+        if len(parts) == 1:
+            user_object = self.context_object
+            xuser_name = user_name
+        else:
+            user_object = self.ui.context[parts[0]]
+            xuser_name = ".".join(parts[1:])
+            user_name = parts[-1]
 
-            if mode in ("from", "both"):
+        if mode in {"from", "both"}:
+            self._bind_from(key, user_object, xuser_name, editor_name, is_list)
 
-                def user_trait_modified(new):
-                    # Need this to include 'user_object' in closure:
-                    user_object
-                    if key not in self._no_trait_update:
-                        self._no_trait_update[key] = None
-                        try:
-                            setattr(self, editor_name, new)
-                        except:
-                            from traitsui.api import raise_to_debug
+            if not is_event:
+                # initialize editor value from user value
+                with self.raise_to_debug():
+                    user_value = xgetattr(user_object, xuser_name)
+                    setattr(self, editor_name, user_value)
 
-                            raise_to_debug()
-                        del self._no_trait_update[key]
+        if mode in {"to", "both"}:
+            self._bind_to(key, user_object, xuser_name, editor_name, is_list)
 
-                user_object.on_trait_change(user_trait_modified, xuser_name)
+            if mode == "to" and not is_event:
+                # initialize user value from editor value
+                with self.raise_to_debug():
+                    editor_value = xgetattr(self, editor_name)
+                    xsetattr(user_object, xuser_name, editor_value)
 
-                if self._user_to is None:
-                    self._user_to = []
-                self._user_to.append(
-                    (user_object, xuser_name, user_trait_modified)
-                )
+    def _bind_from(self, key, user_object, xuser_name, editor_name, is_list):
+        """ Bind trait change handlers from a user object to the editor.
+
+        Parameters
+        ----------
+        key : str
+            The key to use to guard against recursive updates.
+        user_object : object
+            The extended name of the trait to be used on the user object.
+        editor_name : string
```

Comment:
Missing a description of the `xuser_name` parameter, and the description for `user_object` doesn't look right